visualization/app/codeCharta/util/nodeDecorator.ts

```diff
@@ -16,220 +15,210 @@ import { CodeMapHelper } from "./codeMapHelper"
 
-	private static decorateMapWithMissingObjects(map: CodeMapNode) {
-		if (map) {
-			const root = hierarchy<CodeMapNode>(map)
-			root.each(node => {
-				node.data.attributes = !node.data.attributes ? {} : node.data.attributes
-				node.data.edgeAttributes = !node.data.edgeAttributes ? {} : node.data.edgeAttributes
-				Object.assign(node.data.attributes, { [NodeMetricDataService.UNARY_METRIC]: 1 })
-			})
+			// TODO: Verify the need for this code. It is unclear why child
+			// properties are copied to their parent.
+			if (data.children?.length === 1 && data.children[0].children?.length > 0) {
+				const [child] = data.children
+				data.children = child.children
+				data.name += `/${child.name}`
+				data.path += `/${child.name}`
+				if (child.link) {
+					data.link = child.link
+				}
+			}
 		}
 	}
 
-	private static decorateLeavesWithMissingMetrics(map: CodeMapNode, metricData: NodeMetricData[]) {
-		if (map && metricData) {
-			const root = hierarchy<CodeMapNode>(map)
-			root.leaves().forEach(node => {
-				metricData.forEach(metric => {
-					if (node.data.attributes[metric.name] === undefined) {
-						node.data.attributes[metric.name] = 0
-					}
-				})
+	public static decorateMapWithPathAttribute(file: CCFile) {
+		if (file?.map) {
+			const root = hierarchy<CodeMapNode>(file.map)
+			root.each(node => {
+				if (node.parent) {
+					node.data.path = `${node.parent.data.path}/${node.data.name}` 
+				} else {
+					node.data.path = `/${node.data.name}`
+				}
 			})
 		}
+		return file
 	}
 
 	public static decorateParentNodesWithAggregatedAttributes(
 		map: CodeMapNode,
-		blacklist: BlacklistItem[],
 		metricData: NodeMetricData[],
 		edgeMetricData: EdgeMetricData[],
 		isDeltaState: boolean,
 		attributeTypes: AttributeTypes
 	) {
-		if (map) {
-			const root = hierarchy<CodeMapNode>(map)
-			root.each((node: HierarchyNode<CodeMapNode>) => {
-				const leaves: HierarchyNode<CodeMapNode>[] = node.leaves().filter(x => !x.data.isExcluded)
-				this.decorateNodeWithAggregatedChildrenMetrics(leaves, node, metricData, isDeltaState, attributeTypes)
-				this.decorateNodeWithChildrenSumEdgeMetrics(leaves, node, edgeMetricData, attributeTypes)
-			})
-		}
-		return map
-	}
+		const root = hierarchy<CodeMapNode>(map)
+		root.each((node: HierarchyNode<CodeMapNode>) => {
+			let edgeMetrics;
+			const { data } = node
 
-	private static decorateNodeWithAggregatedChildrenMetrics(
-		leaves: HierarchyNode<CodeMapNode>[],
-		node: HierarchyNode<CodeMapNode>,
-		metricData: NodeMetricData[],
-		isDeltaState: boolean,
-		attributeTypes: AttributeTypes
-	) {
-		metricData.forEach(metric => {
-			if (node.data.children && node.data.children.length > 0) {
-				node.data.attributes[metric.name] = this.aggregateLeafMetric(
-					leaves.map(x => x.data.attributes),
-					metric.name,
+			if (!hasChildren(data)) {
+				return
+			}
+
+			for (const { name } of edgeMetricData) {
+				data.edgeAttributes[name] = { incoming: 0, outgoing: 0 }
+			}
```

Comment:
This seems out of place. I'd expect these to be set on the `decorateMap` method already, since we add empty objects there.

Reply:
That would require to pass through the `edgeMetricData` to `decorateMap` and it would for now also increase the amount of work to determine the number correct metrics. There is potential to optimize this further as soon as `decoarateMap` and `decorateParentNodesWithAggregatedAttributes` is combined into a single function. I would not want to do that right now though.